### PR TITLE
Refactor Fact and CanSay datastructure

### PR DESCRIPTION
**Pull request by harsha-mandadi**
1. In this PR, Fact is modified to a list of CanSay and a Base Fact.
2. Accordingly, modified loweing_ast_datalog and souffle emitter tests.
3. Changed the ast_construction tests to have a pretty printer for AST nodes.

### CODE REVIEW

File: src/ir/auth_logic/ast_construction_test.cc [hunk 1]
```diff
@@ -26,51 +26,76 @@
...
+  Principal principal;
+  Fact fact;
 };
 
+std::string ToString(datalog::Predicate predicate) {
+  return absl::StrCat(predicate.name(), "(",
+                      absl::StrJoin(predicate.args(), ", "), ")");
+}
+std::string ToString(BaseFact basefact) {
+  if (basefact.GetValue().index() == 0) {  // Predicate BaseFact
```
Comment (markww) [inline src/ir/auth_logic/ast_construction_test.cc:51]: Don't use `index` on a `std::variant`. Imagine someone removes or adds a variant from the front of the list: all of your indices will get messed up and mean something different, and you'll have to fix every index in the program.

Instead, you should just use `visit` with a lambda to get these alternatives. If every `Datalog` AST structure must have a `ToPrettyString` method, you can just do this:

```C++
// We are within the method `ToPrettyString` in `BaseFact`, so we have access to value_.
std::visit(value_, [](const auto &arg) { arg.ToPrettyString(); }
```

File: src/ir/auth_logic/lowering_ast_datalog.cc [hunk 2]
```diff
@@ -158,58 +164,49 @@ LoweringToDatalogPass::BaseFactToDLIR(const Principal& speaker,
...
+    // ```
+
+    // This is p says inner_fact_dlir
+    auto lhs = PushPrincipal("says_", speaker, inner_fact_dlir);
+    auto fresh_prin_says_inner =
+        PushPrincipal("says_", fresh_principal, inner_fact_dlir);
+    auto speaker_says_fresh_cansay_inner = PushPrincipal(
+        "says_", speaker,
+        PushPrincipal("canSay_", fresh_principal, inner_fact_dlir));
+    auto rhs = {fresh_prin_says_inner, speaker_says_fresh_cansay_inner};
```
Comment (markww) [inline src/ir/auth_logic/lowering_ast_datalog.cc:185]: What is the type of `rhs`? I genuinely don't know. I think `auto` should be replaced with an explicit type annotation here.

File: src/ir/auth_logic/souffle_emitter_test.cc [hunk 3]
```diff
@@ -133,15 +133,13 @@ grounded_dummy(dummy_var).
...
-  Fact::FactVariantType inner_cansay =
-      std::unique_ptr<CanSay>(new CanSay(Principal("PrincipalA"), inner_fact));
-  std::unique_ptr<Fact> inner_cansay_fact =
-      std::unique_ptr<Fact>(new Fact(std::move(inner_cansay)));
-  Fact::FactVariantType cansay_fact = std::unique_ptr<CanSay>(
-      new CanSay(Principal("PrincipalB"), inner_cansay_fact));
+  BaseFact inner_base_fact(
+      datalog::Predicate("grantAccess", {"secretFile"}, datalog::kPositive));
+  // Fact inner_cansay({CanSay(Principal("PrincipalA"))}, inner_base_fact);
+  // Fact inner_cansay_fact(Fact(std::move(inner_cansay)));
```
Comment (markww) [inline src/ir/auth_logic/souffle_emitter_test.cc:139]: Should these be removed?

File: src/ir/auth_logic/ast.h [hunk 4]
```diff
@@ -93,32 +93,25 @@ class BaseFact {
...
-  explicit Fact(FactVariantType value) : value_(std::move(value)) {}
-  const FactVariantType& GetValue() const { return value_; }
+  Fact(std::forward_list<CanSay> cansay, BaseFact basefact)
+      : cansay_(std::move(cansay)), basefact_(std::move(basefact)) {}
+  const std::forward_list<CanSay>& cansay() const { return cansay_; }
+  const BaseFact& basefact() const { return basefact_; }
 
  private:
-  FactVariantType value_;
+  std::forward_list<CanSay> cansay_;
```
Comment (markww) [inline src/ir/auth_logic/ast.h:113]: In general, in C++, `std::vector` is preferred over `std::forward_list`. The reason for this is that usually they are about equally easy to work with, but `std::vector` allocates its memory in one big, contiguous chunk, which is friendlier to caches and memory management.

In this case, however, you've got a pretty good simplicity argument for using `std::forward_list`, as you wish to push elements onto the front of your list. Just wanted to point out that for the future, though.

File: src/ir/auth_logic/ast.h [hunk 5]
```diff
@@ -93,32 +93,25 @@ class BaseFact {
...
+  BaseFact basefact_;
```
Comment (markww) [inline src/ir/auth_logic/ast.h:114]: `basefact_` should be `base_fact_`, as anyplace where you would use CapsToSeparateWords in a caps-case situation like the type name, you should use underscores to separate words in a field or local variable name.

File: src/ir/auth_logic/ast.h [hunk 6]
```diff
@@ -93,32 +93,25 @@ class BaseFact {
...
-  // FactVariantType gives the different forms of Facts. Client code
-  // should use this type to traverse these forms. This type may be changed in
-  // the future.
-  using FactVariantType = std::variant<BaseFact, std::unique_ptr<CanSay>>;
-  explicit Fact(FactVariantType value) : value_(std::move(value)) {}
-  const FactVariantType& GetValue() const { return value_; }
+  Fact(std::forward_list<CanSay> cansay, BaseFact basefact)
+      : cansay_(std::move(cansay)), basefact_(std::move(basefact)) {}
+  const std::forward_list<CanSay>& cansay() const { return cansay_; }
+  const BaseFact& basefact() const { return basefact_; }
```
Comment (markww) [inline src/ir/auth_logic/ast.h:110]: `cansay`, `cansay_` -> `can_say`, `can_say_`,
`basefact`, `basefact_` -> `base_fact`, `base_fact_`.

File: src/ir/auth_logic/ast.h [hunk 7]
```diff
@@ -93,32 +93,25 @@ class BaseFact {
...
-  explicit CanSay(Principal principal, std::unique_ptr<Fact>& fact)
-      : principal_(principal), fact_(std::move(fact)) {}
+  explicit CanSay(Principal principal) : principal_(std::move(principal)) {}
   const Principal& principal() const { return principal_; }
-  const Fact* fact() const { return fact_.get(); }
 
  private:
   Principal principal_;
-  std::unique_ptr<Fact> fact_;
 };
```
Comment (markww) [inline src/ir/auth_logic/ast.h:103]: `CanSay` is looking pretty thin. Would it make more sense for `Fact` to just contain a list of `Principal`s with the list name indicating that they are `CanSay`?

File: src/ir/auth_logic/ast_construction_test.cc [hunk 8]
```diff
@@ -26,51 +26,76 @@
...
+    return absl::StrCat(canactas.left_principal().name(), " canActAs ",
+                        canactas.right_principal().name());
+  }
+}
+std::string ToString(Fact fact) {
+  std::string cansay_string = "";
+  for (const auto& cansay : fact.cansay()) {
+    cansay_string =
+        absl::StrCat(cansay.principal().name(), " canSay ", cansay_string);
+  }
```
Comment (markww) [inline src/ir/auth_logic/ast_construction_test.cc:69]: Don't repeatedly `StrCat` in a loop like this, especially not to the front of a string. This is likely to be quadratic in runtime. Use `StrJoin` instead.

File: src/ir/auth_logic/ast_construction_test.cc [hunk 9]
```diff
@@ -26,51 +26,76 @@
...
 };
 
+std::string ToString(datalog::Predicate predicate) {
+  return absl::StrCat(predicate.name(), "(",
+                      absl::StrJoin(predicate.args(), ", "), ")");
+}
+std::string ToString(BaseFact basefact) {
+  if (basefact.GetValue().index() == 0) {  // Predicate BaseFact
+    return absl::StrCat(
+        ToString(std::get<datalog::Predicate>(basefact.GetValue())));
```
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction_test.cc]: My bad. I was making changes to my original draft to make it look pretty and missed removing absl::StrCat here. Removed it now.

File: src/ir/auth_logic/ast_construction.cc [hunk 10]
```diff
@@ -78,22 +79,27 @@ static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
...
 }
 
-static Fact ConstructFact(AuthLogicParser::FactContext& context) {
+static Fact ConstructFact(std::forward_list<CanSay>& can_say,
+                          AuthLogicParser::FactContext& context) {
   if (auto* flat_fact_fact_ctx =
           dynamic_cast<AuthLogicParser::FlatFactFactContext*>(
               &context)) {  // FlatFactFactContext
     return Fact(
+        std::move(can_say),
```
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction.cc:88]: Done!

File: src/ir/auth_logic/ast_construction.cc [hunk 11]
```diff
@@ -78,22 +79,27 @@ static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
   return BaseFact(ConstructPredicate(*CHECK_NOTNULL(ctx.predicate())));
 }
 
-static Fact ConstructFact(AuthLogicParser::FactContext& context) {
+static Fact ConstructFact(std::forward_list<CanSay>& can_say,
+                          AuthLogicParser::FactContext& context) {
```
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction.cc]: Done!. Made changes to pass by value forward_list

File: src/ir/auth_logic/lowering_ast_datalog.cc (hunk 2 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/lowering_ast_datalog.cc]: Added types. rhs is of type datalog::Predicate

File: src/ir/auth_logic/ast.h (hunk 4 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h]: Noted!

File: src/ir/auth_logic/ast.h (hunk 5 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h]: Done! Resolving similar comments below!

File: src/ir/auth_logic/ast_construction_test.cc (hunk 8 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction_test.cc]: StrJoin needs a separator which in our case is an empty string. This change also might make code not look as redundant.

File: src/ir/auth_logic/ast_construction_test.cc (hunk 1 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction_test.cc]: Removed the usage of index and added std::visit(

File: src/ir/auth_logic/ast_construction_test.cc [hunk 12]
```diff
@@ -26,51 +26,76 @@
...
-  std::string fact_name;
-  datalog::Predicate predicate;
+  Principal principal;
+  Fact fact;
 };
 
+std::string ToString(datalog::Predicate predicate) {
+  return absl::StrCat(predicate.name(), "(",
+                      absl::StrJoin(predicate.args(), ", "), ")");
+}
```
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction_test.cc:50]: Noted! I will take care of these in datalog namespace in another PR.

File: src/ir/auth_logic/ast.h (hunk 7 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h:103]: I feel CanSay will be easy to understand that Fact is a combination of CanSay facts and Base Facts. Removing canSay and just keeping Principal would need someone new to look at grammar to understand Fact.

File: src/ir/auth_logic/ast.h [hunk 13]
```diff
@@ -93,32 +93,25 @@ class BaseFact {
...
-// <principal> canSay <Fact>
 class Fact {
  public:
-  // FactVariantType gives the different forms of Facts. Client code
-  // should use this type to traverse these forms. This type may be changed in
-  // the future.
-  using FactVariantType = std::variant<BaseFact, std::unique_ptr<CanSay>>;
-  explicit Fact(FactVariantType value) : value_(std::move(value)) {}
-  const FactVariantType& GetValue() const { return value_; }
+  Fact(std::forward_list<CanSay> can_say, BaseFact base_fact)
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:107]: After your changes,`Fact` is probably not the right name for this class, but let us leave this as is for now to avoid too many changes.

File: src/ir/auth_logic/ast.h [hunk 14]
```diff
@@ -93,32 +93,25 @@ class BaseFact {
...
+      : can_say_(std::move(can_say)), base_fact_(std::move(base_fact)) {}
+  const std::forward_list<CanSay>& can_say() const { return can_say_; }
+  const BaseFact& base_fact() const { return base_fact_; }
 
  private:
-  FactVariantType value_;
+  std::forward_list<CanSay> can_say_;
```
Comment (bgogul) [inline src/ir/auth_logic/ast.h:113]: If you get rid of `CanSay`, then you can change this to `std::forward_list<Principal> delegatees_;` or `std::forward_list<Principal> delegation_chain_;`?

File: src/ir/auth_logic/ast_construction_test.cc [hunk 15]
```diff
@@ -26,51 +26,78 @@
...
   std::string query_name;
-  std::string principal_name;
-  std::string fact_name;
-  datalog::Predicate predicate;
+  Principal principal;
+  Fact fact;
 };
 
+std::string ToString(datalog::Predicate predicate) {
+  return absl::StrCat(predicate.name(), "(",
```
Comment (bgogul) [inline src/ir/auth_logic/ast_construction_test.cc:48]: How about the sign?

File: src/ir/auth_logic/ast_construction_test.cc [hunk 16]
```diff
@@ -26,51 +26,78 @@
...
-                                datalog::Sign::kNegated)})));
+            {R"(may_notifierA = query "UserA" says mayA(NotifierA))",
+             "may_notifierA", Principal(R"("UserA")"),
+             Fact({}, BaseFact(datalog::Predicate("mayA", {"NotifierA"},
+                                                  datalog::Sign::kPositive)))}),
+        QueryTestData({R"(may_notifierB = query "UserB" says !mayB(NotifierB))",
+                       "may_notifierB", Principal(R"("UserB")"),
+                       Fact({}, BaseFact(datalog::Predicate(
+                                    "mayB", {"NotifierB"},
+                                    datalog::Sign ::kNegated)))})));
```
Comment (bgogul) [inline src/ir/auth_logic/ast_construction_test.cc:100]: Instead of creating a `Fact`, simply add the string representation as an expectation for predicate? e.g., 
```C++
QueryTestData(
  {R"(may_notifierA = query "UserA" says mayA(NotifierA))",
    "may_notifierA",  "UserA",  R"(mayA("NotifierA"))"},
QueryTestData(
  {R"(may_notifierB = query "UserB" says !mayB(NotifierB))",
   "may_notifierB", Principal(R"("UserB")"), R("!mayB("NotifierB")")}));
```

File: src/ir/auth_logic/lowering_ast_datalog.cc [hunk 17]
```diff
@@ -133,17 +139,17 @@ LoweringToDatalogPass::BaseFactToDLIRInner(
 std::pair<datalog::Predicate, std::vector<datalog::Rule>>
 LoweringToDatalogPass::BaseFactToDLIRInner(const Principal& speaker,
                                            const Attribute& attribute) {
-  return std::make_pair(AttributeToDLIR(attribute),
-                        std::vector<datalog::Rule>(
-                            {SpokenAttributeToDLIR(speaker, attribute)}));
+  return std::make_pair(
+      AttributeToDLIR(attribute),
```
Comment (bgogul) [inline src/ir/auth_logic/lowering_ast_datalog.cc:143]: nit: preferable to keep formatting changes in a separate PR.

**Review by bgogul — COMMENTED** [review]
Looks good modulo getting rid of `CanSay` completely.

File: src/ir/auth_logic/ast.h (hunk 7 above)
Comment (aferr) [reply src/ir/auth_logic/ast.h:103]: +1 to just having a list of `Principal`s in `Fact` and getting rid of the `CanSay` class.

File: src/ir/auth_logic/ast_construction_test.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (aferr) [reply src/ir/auth_logic/ast_construction_test.cc:50]: It's probably useful to have `ToPrettyString` methods for the classes in `.../auth_logic/ast.h` in addition to the ones in `src/ir/datalog/program.h`. (I also think `src/ir/auth_logic/souffle_emitter.h` does essentially what `ToPrettyString` for `datalog/program.h` would do, but that could be refactored into methods for `program.h` if desired).

File: src/ir/auth_logic/lowering_ast_datalog.cc (hunk 17 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/lowering_ast_datalog.cc:143]: These format changes were unintentional!! Got auto-formatted on save.

File: src/ir/auth_logic/ast_construction_test.cc (hunk 8 above)
Comment (markww) [reply src/ir/auth_logic/ast_construction_test.cc]: It looks to me like your separator is `" canSay "`

File: src/ir/auth_logic/ast_construction_test.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction_test.cc]: Done!

File: src/ir/auth_logic/ast.h (hunk 7 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h]: Done!

File: src/ir/auth_logic/ast.h (hunk 13 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h]: Resolving this for now. Keeping a note of this in this issue. #532

File: src/ir/auth_logic/ast.h (hunk 14 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast.h]: Done!

File: src/ir/auth_logic/ast_construction_test.cc (hunk 15 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction_test.cc]: Oops! Missed it. Good catch!

File: src/ir/auth_logic/ast_construction_test.cc (hunk 8 above)
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction_test.cc]: Changed to StrJoin wherever possible.

File: src/ir/auth_logic/ast_construction_test.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (harsha-mandadi) [reply src/ir/auth_logic/ast_construction_test.cc]: Keeping track of this in issue#533

File: src/ir/auth_logic/ast.h [hunk 18]
```diff
@@ -89,36 +89,21 @@ class BaseFact {
...
-  using FactVariantType = std::variant<BaseFact, std::unique_ptr<CanSay>>;
-  explicit Fact(FactVariantType value) : value_(std::move(value)) {}
-  const FactVariantType& GetValue() const { return value_; }
+  Fact(std::forward_list<Principal> delegation_chain, BaseFact base_fact)
+      : delegation_chain_(std::move(delegation_chain)),
+        base_fact_(std::move(base_fact)) {}
+  const std::forward_list<Principal>& delegation_chain() const {
+    return delegation_chain_;
+  }
+  const BaseFact& base_fact() const { return base_fact_; }
```
Comment (markww) [inline src/ir/auth_logic/ast.h:102]: Nit: add a line break between these methods.

File: src/ir/auth_logic/ast_construction.cc [hunk 19]
```diff
@@ -78,22 +79,27 @@ static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
...
     return Fact(
+        std::move(delegation_chain),
         ConstructFlatFact(*CHECK_NOTNULL(flat_fact_fact_ctx->flatFact())));
   }
   auto& ctx = *CHECK_NOTNULL(
       dynamic_cast<AuthLogicParser::CanSayFactContext*>(&context));
-  return ConstructFact(*CHECK_NOTNULL(ctx.fact()));
+  delegation_chain.push_front(
+      ConstructPrincipal(*CHECK_NOTNULL(ctx.principal())));
+  return ConstructFact(delegation_chain, *CHECK_NOTNULL(ctx.fact()));
```
Comment (markww) [inline src/ir/auth_logic/ast_construction.cc:95]: `delegation_chain` should be `std::move`'d into `ConstructFact` now that `delegation_chain` is a value type. That allows `ConstructFact` to just copy elide `delegation_chain` into `ConstructFact`'s parameter.

File: src/ir/auth_logic/ast_construction.cc [hunk 20]
```diff
@@ -78,22 +79,27 @@ static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
...
 }
 
 static Query ConstructQuery(AuthLogicParser::QueryContext& context) {
+  std::forward_list<Principal> delegation_chain;
   return Query(context.ID()->getText(),
                ConstructPrincipal(*CHECK_NOTNULL(context.principal())),
-               ConstructFact(*CHECK_NOTNULL(context.fact())));
+               ConstructFact(delegation_chain, *CHECK_NOTNULL(context.fact())));
```
Comment (markww) [inline src/ir/auth_logic/ast_construction.cc:102]: Now, `delecation_chain` should be moved at this callsite. Alternatively, you can just indicate that you want an initial empty list with an initializer:

```C++
ConstructFact({}, *CHECK_NOTNULL(context.fact()));
```

**Review by markww — APPROVED** [review]
Looks good modulo some comments.

**Review by bgogul — APPROVED** [review]
